File: Week2/Python/ArrayProblem/MissingNumber.py
"""
    Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range 
    that is missing from the array.

    https://leetcode.com/problems/missing-number/
"""


def missingNumber(nums):
    """ T(c) -> O(n) , S(c) -> O(1) """
    n = len(nums)
    current_total = 0
        # total will contain sum of numbers range from 0 to n
    total = (n * (n + 1)) / 2
    
    for i in range(len(nums)):
        # current_total will contain sum of all numbers inside nums array
        current_total += nums[i]
    
    missing_num = total - current_total
    
    return missing_num


arr = [9,6,4,2,3,5,7,0,1]
print(missingNumber(arr))

File: Week2/Java/ArrayProblem/MissingNumber.java
public class MissingNumber {
/*
* Given an array nums containing n distinct numbers in the range [0, n], return the only number in the range
* that is missing from the array.
* https://leetcode.com/problems/missing-number/
* T(c) -> O(n), S(c) -> O(1)
*/    
    static int missingNumber(int[] nums) {
        int n = nums.length;
        int naturalSum = (n*(n+1))/2;
        
        int actualSum = 0;
        for(int i=0;i<n;i++)
        {
            actualSum += nums[i];
        }
        
        int missingNum = naturalSum - actualSum;
        return missingNum;
    } 

    public static void main(String[] args) {
        int nums[] = {9,6,4,2,3,5,7,0,1};
        System.out.println(missingNumber(nums));
    }
}
